### PR TITLE
Closes #3527

### LINKED ISSUE #3527: Prefer if over match simple sequences and mappings
# Rule request

## Thesis
We should support simplifying `match` statements that use simple sequence and mapping patterns:

Example:
```python
match data:
    case [1, 2]:
        handle_pair()
    case _:
        ignore()
```
Can be rewritten as:
```python
if data == [1, 2]:
    handle_pair()
else:
    ignore()
```

## Reasoning
Using match for exact structural comparisons of simple literals is unnecessarily verbose. While match excels at deconstruction, using it to check for an exact list or dict value is better expressed with a direct equality comparison (`==`), which is more readable and performant.

These patterns are common in configuration checks, API responses, and state validation.

When should this be allowed?

Only when:
- There are exactly two cases (with `case _:`)
- The first pattern is a simple sequence like `[1, "a"]` or a simple mapping like `{"x": 1}`
- All elements/keys/values are simple (literals, constants, names)
- No starred patterns (`*rest`) or variable bindings (`case [x, y]:`)
- Keys in mappings are constants (not variables)

Not allowed if:
- Uses deconstruction (e.g. `case [x, y]:`)
- Contains `*args` or *`*kwargs`
- Has guards (`if` ...)

### PR DESCRIPTION
**Pull request by Khabib73**
# I have made things!

## Checklist

- [x] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [x] I have created at least one test case for the changes I have made
- [x] I have updated the documentation for the changes I have made
- [x] I have added my changes to the `CHANGELOG.md`

## Related issues

Closes #3527

🙏 Please, if you or your company is finding wemake-python-styleguide valuable, help us sustain the project by sponsoring it transparently on https://opencollective.com/wemake-python-styleguide. As a thank you, your profile/company logo will be added to our main README which receives hundreds of unique visitors per day.

### CODE REVIEW

File: tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py [hunk 1]
```diff
@@ -115,6 +108,214 @@
...
+match subject:
+    case [True, 1, 'a', False]:
+        pass
+    case _:
+        pass
+"""
+
+sequences_with_constants = """
+match subject:
+    case [some.CONSTANT]:
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py]: Sorry, this does not seem correct. How can we tell that this is a constant?

File: tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py [hunk 2]
```diff
@@ -115,6 +108,214 @@
...
+match subject:
+    case [True, 1, False]:
+        pass
+    case _:
+        pass
+"""
+
+sequences_with_bools_and_strings = """
+match subject:
+    case [True, 'a', False]:
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py]: Please, also test `None`

File: tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py [hunk 3]
```diff
@@ -115,6 +108,214 @@
...
+match subject:
+    case [True, False]:
+        pass
+    case _:
+        pass
+"""
+
+sequences_with_nums_and_strings = """
+match subject:
+    case [1, 'a', 2]:
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py]: Maybe we can format this one template with different patterns?

File: tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py [hunk 4]
```diff
@@ -115,6 +108,214 @@
...
+        sequences_with_bools,
+        sequences_with_constants,
+        sequences_with_nums_and_strings,
+        sequences_with_bools_and_nums,
+        sequences_with_bools_and_strings,
+        sequences_with_bools_and_nums_and_strings,
+    ],
+)
+def test_simplifiable_sequence(
+    assert_errors, parse_ast_tree, default_options, code
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py]:
```suggestion
    assert_errors, parse_ast_tree, default_options, code,
```

nit

you can enforce this rule in a separate PR, btw: see `COM812`

Comment (Khabib73) [reply tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py]: @sobolevn ruff format conflicting with ruff checking:

```
"COM812", # trailing comma, conflicts with `ruff format`
```

Seems like this problem is still not fixed. Or do you mean to write own rule with WPS?

File: tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py (hunk 3 above)
Comment (Khabib73) [reply tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py]: sure!

File: tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py (hunk 4 above)
Comment (sobolevn) [reply tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py]: Yes, this is fine :)
We use `COM812` everyone in our projects (because this is how it should be by default).

**Review by sobolevn — APPROVED** [review]
Thanks!

File: tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py (hunk 4 above)
Comment (sobolevn) [reply tests/test_visitors/test_ast/test_conditions/test_simplified_match_conditions.py]: You can do that in a separate PR :)
